### ISSUE TITLE
Basemap accepted endpoints

### ISSUE THREAD

**Issue by prosperva** [fgpv-vpgf/fgpv-vpgf#4076]
What type of endpoints are accepted to add a base map. I tried adding an endpoint that returns an image with `image/jpeg` and `image/png` but I get this error: 

`Refused to execute script from 'https://maps.geogratis.gc.ca/wms/canvec_en?SERVICE=wms&REQUEST=GetMap&VERSION=1.3.0&LAYERS=transport&CRS=EPSG:4269&BBOX=37,-142,90,-48&FORMAT=image/png&WIDTH=3408&HEIGHT=2394&f=json&callback=dojo_request_script_callbacks.dojo_request_script0' because its MIME type ('image/png') is not executable`

**Comment by james-rae** [comment]
The basemap stuff is leveraging the basemap management system that comes in the ESRI API (v3). This gives us a lot of stuff for free (able to swap out basemaps with a few lines of code, etc).

The RAMP config structure for basemaps is a simplification on top of that; specifically you can specify the `url` and `opacity` and that's about it. There is a `layerType` in the schema but far as I can tell right now it's not being used (believe it was a very long time ago, but the schema is not very helpful :( in explaining it or indicating it's obsolete ).

All that said, the [ESRI API BasemapLayer class documentation](https://developers.arcgis.com/javascript/3/jsapi/basemaplayer-amd.html) indicates that for a URL sourced basemap, you are limited to something on an ArcGIS Server that returns a map (tile services, dynamic layer map, ImageServer likely works as well). A WMS would not be supported.

**Comment by prosperva** [comment]
Thanks @james-rae

**Comment by teaganryley** [comment]
> All that said, the ESRI API BasemapLayer class documentation indicates that for a URL sourced basemap, you are limited to something on an ArcGIS Server that returns a map (tile services, dynamic layer map, ImageServer likely works as well). A WMS would not be supported.

@james-rae does this apply for other layer types as well? For example, suppose I wanted to add a single WMS layer that was not part of a basemap.

**Comment by james-rae** [comment]
No, regular layers are not spun up using the `BasemapLayer` class. 

We do have limited™ support for WMS layers, the fancier aspects of those services may not be supported or fail in spectacular fashion.

As an example, the default layer on [this site](https://climate-viewer.canada.ca/climate-maps.html) has a layer config like this (I've removed some cruft)

```json
{
  "id": "DCS_tmax_ANN_rcp85_2081_en",
  "layerType": "ogcWms",
  "suppressGetCapabilities": true,
  "url": "https://geo.weather.gc.ca/geomet-climate?SERVICE=WMS&VERSION=1.3.0",
  "name": "Daily maximum temperature",
  "state": {
    "opacity": 0.85,
    "visibility": true
  },
  "layerEntries": [
    {
      "id": "DCS.TX.RCP85.YEAR.2081-2100_PCTL50"
    }
  ],
  "featureInfoMimeType": "application/json"
}
```

Note the `suppressGetCapabilities` flag is mainly for this geomet service, to avoid downloading the massive getcapabilities response it likes to serve.

**Comment by teaganryley** [comment]
Thank you again @james-rae, your explanation is greatly appreciated. This brings me to two questions:

In the example above, your WMS request specifies only two parameters (service and version). Was the URL truncated for the sake of example, or is this the proper way to point to a WMS service?

Secondly, the climate viewer map has some features that I've been looking to implement in RAMP. Would it be possible to get the source code?

**Comment by james-rae** [comment]
That sample url is un-modified. The value from the config gets passed to an internal ESRI [WMSLayer](https://developers.arcgis.com/javascript/3/jsapi/wmslayer-amd.html) class. Given how (at least in my experience) WMS url structures can vary based on server, and ESRI's own limited documentation on the url property (it looks like they just cut & pasted the generic text to be honest), there is often some trial and error to get a WMS url format to behave itself. 

A faster way to experiment may be just using [this esri sandbox sample](https://developers.arcgis.com/javascript/3/sandbox/sandbox.html?sample=layers_wmsresourceinfo), remove the hardcoded `resourceInfo` stuff, and fiddle around until your layer draws on the map.

But generally, the url supplied is consumed by the ESRI API and they have internal code that will leverage the url to structure the `GetCapabilities` call (which happens on layer load) and the `GetMap` call upon any redraw requests.

For that site in particular, the source is unfortunately on a private repo behind a firewall.

Most of the fancy stuff you see is custom UI & code that sits on the website and just interacts with the RAMP viewer via the API (e.g. the time slider / horizontal legend, the data selector along the top). Not even using the same js framework as RAMP.  The custom details panels are using the [templating framework](https://fgpv-vpgf.github.io/fgpv-vpgf/develop/docs/#/mapauthor/details_templating). 

If there is something else in particular that has your interest, reply and I can see if I can shed more light. Note I won't be back online until Wednesday.

**Comment by teaganryley** [comment]
Thank you again for all your help thus far. To make sure I understand things correctly:

This component is custom UI that sits outside of RAMP, correct?
![climate-ui](https://user-images.githubusercontent.com/29206084/205079818-95e1a90b-e74b-4cd5-a453-beb0bcd0e188.png)

And this component is actually one of the RAMP plugins (thematic slider I think)?
![climate-slider](https://user-images.githubusercontent.com/29206084/205080092-3a8b6720-b09c-4022-bb50-366777dc6ee7.png)

I didn't understand which part involved the templating framework. Could you please clarify?

In general, I'm interested in two problems:
1. Sending data to RAMP, from outside (example: A dropdown menu for a location triggers panning/zooming in RAMP)
2. Getting data *out* of RAMP (example: RAMP sends the current extent to a UI component outside of RAMP)

The first problem seems quite solvable (the Climate Viewer is doing exactly that), but it's not clear to me how to approach the second problem. Any suggestions you have would be greatly appreciated.

**Comment by james-rae** [comment]
The slider is also custom UI. It sits on top of the map control. Our needs were a bit unique for the site, thematic slider would not do what we required.

For the custom templates, if you visit the site and click anywhere on the temperature heatmap, you'll see the content of the details pane is customized instead of the stock text blob or attributes list.

Extent can be manipulated via the map API.
You can grab your api handle with something like
`var mapAPI = RAMP.mapById('<your map id>');`

You will see some stuff on that named `bounds`, but I advise not using those since they always get translated to Lat/Long projection and thus are warped.

I suggest using
```text
var currExtent = mapAPI.getExtent();
mapAPI.setExtent(newExtent, fit); // fit is optional, set to true to force the entire extent to be in view
```

The object format for the extents is in line with the [common ESRI standard](https://developers.arcgis.com/rest/services-reference/enterprise/extent-ra-.htm), i.e. four number props and a spatial reference subobject.

**Comment by teaganryley** [comment]
Thank you very much!
